src/config/enterprise.rs

```diff
@@ -0,0 +1,426 @@
+    // that Vector was spawned with.
+    let (table, _) = process_paths(config_paths)
+        .map(|paths| load_source_from_paths(&paths).ok())
+        .flatten()
+        .expect("Couldn't load source from config paths. Please report.");
+
+    // Set the relevant fields needed to report a config to Datadog. This is a struct rather than
+    // exploding as func arguments to avoid confusion with multiple &str fields.
+    let fields = PipelinesStrFields {
+        config_version,
+        vector_version: &vector_version,
+    };
+
+    // Set the Datadog authorization fields. There's an API and app key, to allow read/write
+    // access in tandem with RBAC on the Datadog side.
+    let auth = PipelinesAuth {
+        api_key: &api_key,
+        application_key: &datadog.application_key,
+    };
+
+    // Create a HTTP client for posting a Vector version to Datadog OP. This will
+    // respect any proxy settings provided in top-level config.
+    let client = HttpClient::new(None, &datadog.proxy)
+        .expect("couldn't instrument Datadog HTTP client. Please report");
+
+    // Endpoint to report a config to Datadog OP.
+    let endpoint = get_reporting_endpoint(datadog.site.as_ref(), &datadog.configuration_key);
+
+    // Datadog uses a JSON:API, so we'll serialize the config to a JSON
+    let payload = PipelinesVersionPayload::new(&table, &fields);
+
+    // Attempt to report a config to Datadog. This should happen in a loop, up to a maximum
+    // of `max_retries`.
+    for _ in 0..datadog.max_retries {
+        select! {
+            biased;
+            Ok(SignalTo::Shutdown | SignalTo::Quit) = signal_rx.recv() => return Err(PipelinesError::Interrupt),
+            report = report_serialized_config_to_datadog(&client, &endpoint, &auth, &payload) => {
+                match report {
+                    Ok(()) => {
+                        info!(
+                            "Vector config {} successfully reported to {}.",
+                            &config_version, DATADOG_REPORTING_PRODUCT
+                        );
+                        break;
+                    }
+                    Err(err) => {
+                        error!(
+                            err = ?err.to_string(),
+                            "Could not report Vector config to {}.", DATADOG_REPORTING_PRODUCT
+                        );
+
+                        if let ReportingError::StatusCode(status) = err {
+                            // If the error is 'fatal', and the user has elected to exit on fatal errors,
+                            // return control back upstream.
+                            //
+                            // A fatal error is currently determined to be one of the following:
+                            //
+                            // - 4xx class of status code (429 excepted; treated as a retry).
+                            if status.is_client_error() && status != StatusCode::TOO_MANY_REQUESTS {
```

Comment:
All 'client' errors are deemed fatal (since the payload is never going to change).

The exception is `429`, which is interpreted to mean a retry.